mantle/platform/metal.go

```diff
@@ -0,0 +1,450 @@
+var baseKargs = []string{"rd.neednet=1", "ip=dhcp"}
+var liveKargs = []string{"ignition.firstboot", "ignition.platform.id=metal"}
+
+var (
+	// TODO expose this as an API that can be used by cosa too
+	consoleKernelArgument = map[string]string{
+		"amd64-usr":   "ttyS0",
+		"ppc64le-usr": "hvc0",
+		"arm64-usr":   "ttyAMA0",
+		"s390x-usr":   "ttysclp0",
+	}
+)
+
+type Install struct {
+	CosaBuildDir string
+	CosaBuild    *cosa.Build
+
+	Board    string
+	Firmware string
+	Insecure bool
+	QemuArgs []string
+
+	LegacyInstaller bool
+
+	// These are set by the install path
+	kargs    []string
+	ignition string
+}
+
+type InstalledMachine struct {
+	tempdir  string
+	QemuInst *QemuInstance
+}
+
+func (inst *Install) PXE(kargs []string, ignition string) (*InstalledMachine, error) {
+	if inst.CosaBuild.BuildArtifacts.Metal.Path == "" {
+		return nil, fmt.Errorf("Build %s must have a `metal` artifact", inst.CosaBuild.OstreeVersion)
+	}
+
+	inst.kargs = kargs
+	inst.ignition = ignition
+
+	var err error
+	var mach *InstalledMachine
+	if inst.LegacyInstaller {
+		if inst.CosaBuild.BuildArtifacts.Kernel.Path == "" {
+			return nil, fmt.Errorf("build %s has no legacy installer kernel", inst.CosaBuild.OstreeVersion)
+		}
+		mach, err = inst.runLegacy(&kernelSetup{
+			kernel:    inst.CosaBuild.BuildArtifacts.Kernel.Path,
+			initramfs: inst.CosaBuild.BuildArtifacts.Initramfs.Path,
+		})
+		if err != nil {
+			return nil, errors.Wrapf(err, "legacy installer")
+		}
+	} else {
+		if inst.CosaBuild.BuildArtifacts.LiveKernel.Path == "" {
+			return nil, fmt.Errorf("build %s has no live installer kernel", inst.CosaBuild.Name)
+		}
+		mach, err = inst.runLive(&kernelSetup{
+			kernel:    inst.CosaBuild.BuildArtifacts.LiveKernel.Path,
+			initramfs: inst.CosaBuild.BuildArtifacts.LiveInitramfs.Path,
+		})
+		if err != nil {
+			return nil, errors.Wrapf(err, "testing live installer")
+		}
+	}
+
+	return mach, nil
+}
+
+func (*InstalledMachine) Destroy() {
+}
+
+type kernelSetup struct {
+	kernel, initramfs string
+}
+
+type pxeSetup struct {
+	tftpipaddr    string
+	boottype      string
+	networkdevice string
+	bootindex     string
+	pxeimagepath  string
+
+	// bootfile is initialized later
+	bootfile string
+}
+
+type installerRun struct {
+	inst    *Install
+	builder *QemuBuilder
+
+	builddir string
+	tempdir  string
+	tftpdir  string
+
+	metalimg  string
+	metalname string
+
+	baseurl string
+
+	kern kernelSetup
+	pxe  pxeSetup
+}
+
+func absSymlink(src, dest string) error {
+	src, err := filepath.Abs(src)
+	if err != nil {
+		return err
+	}
+	return os.Symlink(src, dest)
+}
+
+func (inst *Install) setupTftpDir(builddir, tftpdir, metalimg string, kern *kernelSetup) error {
+	serializedConfig := []byte(inst.ignition)
+	if err := ioutil.WriteFile(filepath.Join(tftpdir, "config.ign"), serializedConfig, 0644); err != nil {
+		return err
+	}
+
+	metalIsCompressed := !strings.HasSuffix(metalimg, ".raw")
```

Comment:
Given the code in `setup` is this something that could ever realistically be `false`?

Reply:
This is the code that compresses it to start if it's not already compressed.

Reply:
Yes, but the code in `setup` just adds `.gz` to the suffix without checking if it's already compressed or not. Won't `metalimg` never have a `.raw` suffix when the code reaches this point?

Reply:
It is confusing because there are two separate variables for this reason - `metalimg` and `metalname`.  `setup` only changes `metalname` - what we refer to in URLs.